### PR TITLE
Fix #397: unique site constraint failure

### LINKED ISSUE #397: Crash report 7.0-rc-4: Unique site constraint
We should probably convert this into a `DuplicateSiteException`.
```
Caused by android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: SiteModel.SITE_ID, SiteModel.URL (Sqlite code 2067), (OS error - 2:No such file or directory)
       at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(SQLiteConnection.java)
       at android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:742)
       at android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:754)
       at android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:64)
       at android.database.sqlite.SQLiteDatabase.updateWithOnConflict(SQLiteDatabase.java:1627)
       at android.database.sqlite.SQLiteDatabase.update(SQLiteDatabase.java:1568)
       at com.yarolegovich.wellsql.UpdateQuery.execute(UpdateQuery.java:79)
       at org.wordpress.android.fluxc.persistence.SiteSqlUtils.insertOrUpdateSite(SiteSqlUtils.java:139)
       at org.wordpress.android.fluxc.store.SiteStore.updateSite(SiteStore.java:745)
       at org.wordpress.android.fluxc.store.SiteStore.onAction(SiteStore.java:612)
       at java.lang.reflect.Method.invoke(Method.java)
       at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
       at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479)
       at org.greenrobot.eventbus.AsyncPoster.run(AsyncPoster.java:46)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
       at java.lang.Thread.run(Thread.java:833)
```

### PR DESCRIPTION
**Pull request by maxme**
Fix #397: unique site constraint failure

### CODE REVIEW

File: example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java [hunk 1]
```diff
@@ -574,4 +574,24 @@ public void testSingleInsertSiteDuplicateWPCom()
         assertEquals(0, res2.rowsAffected);
         assertEquals(1, mSiteStore.getSitesCount());
     }
+
+    @Test
+    public void testUpdateSiteUniqueConstraintFail() throws DuplicateSiteException {
+        // Create 2 test sites
+        SiteModel site1 = generateTestSite(1, "https://pony1.com", "https://pony1.com/xmlrpc.php", false, true);
```
Comment (aforcier) [inline example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java]: These should really have remote id `0` if they're self-hosted sites, for realism.

Comment (maxme) [inline example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java]: Done in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/398/commits/a297c294ecee9f5f618f3fcbaa46907e0b767b8e

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java [hunk 2]
```diff
@@ -135,8 +136,16 @@ public static int insertOrUpdateSite(SiteModel site) throws DuplicateSiteExcepti
...
-                    .put(site, new UpdateAllExceptId<SiteModel>()).execute();
+            try {
+                return WellSql.update(SiteModel.class).whereId(oldId)
+                        .put(site, new UpdateAllExceptId<SiteModel>()).execute();
+            } catch (SQLiteException e) {
+                AppLog.e(T.DB, "Error while updating site: siteId=" + site.getSiteId() + " url=" + site.getUrl()
+                        + " xmlrpc=" + site.getXmlRpcUrl(), e);
+                // Can happen on self hosted sites with incorrect url values in wp.getOption response.
+                // See https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/397
+                throw new DuplicateSiteException();
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java]: I'm still not sure about this 😞 Couldn't we have other, unrelated `SQLiteException`s that we don't want to silently bury here?

**Comment by aforcier** [comment]
Looks good! I wanted to test this under real conditions since Robolectric is using a different SQLite stack in our tests:

1. Modify example app to fetch first two sites [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/ccd9537eaa9809add385839c55b667d21777fde5/example/src/main/java/org/wordpress/android/fluxc/example/SitesFragment.java#L58)
2. Undo this PR's changes to `SiteSqlUtils`
3. Sign in to self-hosted multi-site in modified example app
4. Modify response from API to replace site 2 URL with site 1 URL in `home_url` field of `wp.getOptions`
5. Fetch site --> 💥
6. Restore this PR's changes
7. Fetch site --> 👍

Caught error:
```
 Error while updating site: siteId=0 url=<site1url> xmlrpc=<site2xmlrpc.php>
        android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: SiteModel.SITE_ID, SiteModel.URL (code 2067)
```
And received `DUPLICATE_SITE` error in example app log, no crash.

**Comment by aforcier** [comment]
Just needs some logging WPAndroid-side so we can see these in Crashlytics!

:shipit:

**Pull request merged**